SHOULD NEVER BE REACHED in IPC::StreamClientConnection::trySendDestinationIDIfNeeded

kkinnunen-apple opened the pull request:
#### da8b5ed3e46e77b73d5eea1506ca091774d6c580
<pre>
SHOULD NEVER BE REACHED in IPC::StreamClientConnection::trySendDestinationIDIfNeeded
<a href="https://bugs.webkit.org/show_bug.cgi?id=250318">https://bugs.webkit.org/show_bug.cgi?id=250318</a>
rdar://104027243

Reviewed by Matt Woodrow and Simon Fraser.

The stream IPC write and read both select the next offset
based on if minimumMessageSize can fit to the allocation. If it
does not fit between current position and buffer end, the index
is wrapped to 0.

The minimumMessageSize is currently expected to match the
SetStreamDestinationID message. The message size can vary between
10 and 16 bytes, since the uint64_t in it is encoded with uint64_t
alignment, 8 bytes.

If the SetStreamDestinationID did not fit to the allocation,
the send was returned as failed. The expectation was that callers
would mark the context lost. This doesn&apos;t happen for normal rendering.
Thus if this would happen for example at DisplayListRecorder creation
message, the client would continue sending messages to destinations
that did not exist at the server. In these cases the server does not
know how to parse the messages, and the expectation is that these
cases happen when the connection is being closed.

Adjust the minimumMessageSize to 16 for now. Add tests testing that
the encoding succeeds at any expected offset.

* Source/WebKit/Platform/IPC/StreamClientConnectionBuffer.h:
(IPC::StreamClientConnectionBuffer::StreamClientConnectionBuffer):
* Source/WebKit/Platform/IPC/StreamConnectionEncoder.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionBufferTests.cpp:
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:
(TestWebKitAPI::TEST_P):

Canonical link: <a href="https://commits.webkit.org/259069@main">https://commits.webkit.org/259069@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eaa184a3c9046f2467e489810b63a243d9b1e6be

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103746 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12862 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36689 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112971 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173292 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107696 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13883 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3756 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95998 "Built successfully") | [💥 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112104 "An unexpected error occured. Recent messages:Failed to print configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109517 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10709 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93773 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38430 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92533 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25379 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80079 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6221 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26773 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6395 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3290 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12378 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46294 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6238 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8156 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebKit/Platform/IPC/StreamConnectionEncoder.h:
```diff
@@ -39,8 +39,9 @@ template<typename, typename> struct ArgumentCoder;
-    static constexpr size_t minimumMessageSize = sizeof(MessageName) + sizeof(uint64_t);
+    // Stream allocation needs to be at least size of StreamSetDestinationID message at any offset % messageAlignment.
+    // StreamSetDestinationID has MessageName+uint64_t, where uint64_t is expected to to be aligned at 8.
+    static constexpr size_t minimumMessageSize = 16;
```
Is it possible that we need 17?

MessageName is 2-byte aligned, so we could end up needing 3-bytes to write it. That puts us on a guaranteed 2-byte boundary, so we could need to align another 6, before writing the 8 of uint64_t.

I'd probably prefer this to be written as `(alignof(MessageName) - 1) + sizeof(MessageName) + (alignof(uint64_t) -1) + sizeof(uint64_t)`.

As discussed on Slack, it'd also be nice if `uint64_t` were replaced with something strongly typed, but that's probably out of scope for this PR.

kkinnunen-apple replied:
>Is it possible that we need 17. MessageName is 2-byte aligned, so we could end up needing 3-bytes to write it. 

No, the message allocation itself is currently aligned per messageAlignment, see the constant one line below.

>I'd probably prefer this to be written as (alignof(MessageName) - 1) + sizeof(MessageName) + (alignof(uint64_t) -1) + sizeof(uint64_t).

It's a valid idea, but I thought it was a source of bugs in its own. Your wording about where MessageName ends and possible alignment values for uint64_t was exactly correct, but your formula was not.

When done correctly, the reader needs to think as hard or harder at the formula as they need for the value 16. 

As an anecdotal note, you wrote `1+2+7+8` which is 18 when you thought you wrote 17. :)

But that's a wrong formula, even accounting for the above information. So it's not as easy to get right or to read as you'd think.  `sizeof(MessageName) + (alignof(uint64_t)-1) + sizeof(uint64_t)`  is `2 + 7 + 8 == 17`.

So if we'd simplify and accept that MessageName is always aligned, the generic idea of the formula would be that the uint64_t align factor depends on the message name size mod uint64_t alignment. So something like:
`(sizeof(MessageName) + (alignof(uint64_t) - (sizeof(MessageName) % alignof(uint64_t)) + sizeof(uint64_t)`

Which quite frankly starts to get quite complex for what it accomplishes.

mattwoodrow commented on Source/WebKit/Platform/IPC/StreamClientConnectionBuffer.h:
```diff
@@ -81,6 +81,11 @@ inline StreamClientConnectionBuffer::StreamClientConnectionBuffer(unsigned dataS
     ASSERT(dataSizeLog2 < 31u); // Currently expected to be not that big, and offset to fit in size_t with the tag bits.
+    // Currently the minimum message size is 16, and as such 32 bytes is not enough to hold one message.
+    // The problem happens after initial write and read, because after the read the buffer is split between
+    // 15 free bytes and 15 free bytes.
```
free and free?

kkinnunen-apple committed:
SHOULD NEVER BE REACHED in IPC::StreamClientConnection::trySendDestinationIDIfNeeded
https://bugs.webkit.org/show_bug.cgi?id=250318
rdar://104027243

Reviewed by Matt Woodrow and Simon Fraser.

The stream IPC write and read both select the next offset
based on if minimumMessageSize can fit to the allocation. If it
does not fit between current position and buffer end, the index
is wrapped to 0.

The minimumMessageSize is currently expected to match the
SetStreamDestinationID message. The message size can vary between
10 and 16 bytes, since the uint64_t in it is encoded with uint64_t
alignment, 8 bytes.

If the SetStreamDestinationID did not fit to the allocation,
the send was returned as failed. The expectation was that callers
would mark the context lost. This doesn't happen for normal rendering.
Thus if this would happen for example at DisplayListRecorder creation
message, the client would continue sending messages to destinations
that did not exist at the server. In these cases the server does not
know how to parse the messages, and the expectation is that these
cases happen when the connection is being closed.

Adjust the minimumMessageSize to 16 for now. Add tests testing that
the encoding succeeds at any expected offset.

* Source/WebKit/Platform/IPC/StreamClientConnectionBuffer.h:
(IPC::StreamClientConnectionBuffer::StreamClientConnectionBuffer):
* Source/WebKit/Platform/IPC/StreamConnectionEncoder.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionBufferTests.cpp:
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:
(TestWebKitAPI::TEST_P):

Canonical link: https://commits.webkit.org/259069@main
